daemons/execd/remoted_schemas.c

```diff
@@ -0,0 +1,248 @@
+ *
+ * This source code is licensed under the GNU Lesser General Public License
+ * version 2.1 or later (LGPLv2.1+) WITHOUT ANY WARRANTY.
+ */
+
+#include <crm_internal.h>
+
+#include <ftw.h>
+#include <unistd.h>
+#include <sys/stat.h>
+
+#include <crm/cib.h>
+#include <crm/cib/cib_types.h>
+#include <crm/cib/internal.h>
+#include <crm/crm.h>
+#include <crm/msg_xml.h>
+#include <crm/common/mainloop.h>
+
+#include "pacemaker-execd.h"
+
+static int
+rm_files(const char *pathname, const struct stat *sbuf, int type, struct FTW *ftwb)
+{
+    /* Don't delete PCMK__REMOTE_SCHEMA_DIR . */
+    if (pcmk__str_eq(pathname, PCMK__REMOTE_SCHEMA_DIR, pcmk__str_none)) {
+        return 0;
+    }
+
+    if (remove(pathname) != 0) {
+        int rc = errno;
+        crm_err("Could not remove %s: %s", pathname, pcmk_rc_str(rc));
+        return -1;
+    }
+
+    return 0;
+}
+
+static void
+clean_up_extra_schema_files(void)
+{
+    struct stat sb;
+    int rc;
+
+    rc = stat(PCMK__REMOTE_SCHEMA_DIR, &sb);
+
+    if (rc == -1) {
+        if (errno == ENOENT) {
+            /* If the directory doesn't exist, try to make it first. */
+            if (mkdir(PCMK__REMOTE_SCHEMA_DIR, 0755) != 0) {
+                rc = errno;
+                crm_err("Could not create directory for schemas: %s",
+                        pcmk_rc_str(rc));
+            }
+
+        } else {
+            rc = errno;
+            crm_err("Could not create directory for schemas: %s",
+                    pcmk_rc_str(rc));
+        }
+
+    } else if (!S_ISDIR(sb.st_mode)) {
+        /* If something exists with the same name that's not a directory, that's
+         * an error.
+         */
+        crm_err("%s already exists but is not a directory", PCMK__REMOTE_SCHEMA_DIR);
+
+    } else {
+        /* It's a directory - clear it out so we can download potentially new
+         * schema files.
+         */
+        rc = nftw(PCMK__REMOTE_SCHEMA_DIR, rm_files, 10, FTW_DEPTH|FTW_MOUNT|FTW_PHYS);
+
+        if (rc != 0) {
+            crm_err("Could not remove %s: %s", PCMK__REMOTE_SCHEMA_DIR, pcmk_rc_str(rc));
+        }
+    }
+}
+
+static void
+write_extra_schema_file(xmlNode *xml, void *user_data)
+{
+    const char *file = NULL;
+    char *path = NULL;
+    int rc;
+
+    file = crm_element_value(xml, PCMK__XA_PATH);
+    if (file == NULL) {
+        crm_warn("No destination path given in schema request");
+        return;
+    }
+
+    path = crm_strdup_printf("%s/%s", PCMK__REMOTE_SCHEMA_DIR, file);
+
+    /* The schema is a CDATA node, which is a child of the <file> node.  Traverse
+     * all children and look for the first CDATA child.  There can't be more than
+     * one because we only have one file attribute on the parent.
+     */
+    for (xmlNode *child = xml->children; child != NULL; child = child->next) {
+        FILE *stream = NULL;
+
+        if (child->type != XML_CDATA_SECTION_NODE) {
+            continue;
+        }
+
+        stream = fopen(path, "w+");
```

Comment:
FYI, I believe we set a umask turning off group write and other read/write, which should be fine for anything that needs schemas (even ACL-limited users will be in the HA group).